### PR TITLE
[BEAM-13974] Handle idle Storage Api streams

### PR DESCRIPTION
**Pull request by reuvenlax**

### CODE REVIEW

**Comment by reuvenlax** [comment]
R: @yirutang

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
@yirutang friendly ping

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java [hunk 1]
```diff
@@ -404,22 +428,27 @@ public String toString() {
...
+              // happened after an append but before the worker could checkpoint it's
+              // state. The records that were appended in a failed bundle will be retried,
+              // meaning that the unflushed tail of the stream must be discarded to prevent
+              // duplicates.
+              boolean offsetMismatch =
+                  statusCode.equals(Code.OUT_OF_RANGE) || statusCode.equals(Code.ALREADY_EXISTS);
+              // This implies that the stream doesn't exist or has already been finalized. In this
+              // case we have no
+              // choice but to create a new stream.
+              boolean streamDoesntExist = statusCode.equals(Code.INVALID_ARGUMENT);
```
Comment (baeminbo) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:443]: I guess `FAILED_PRECONDITIONS` can be returned when a WriteStream is finalized by idleness.

Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:443]: Does it? This seems like a bug if it's the case, as FAILED_PRECONDITION is generally used for errors that are fixable by the user (e.g. creating a file in a directory that does not exists), while this is an error that will always continue regardless of what the user does.

**Comment by reuvenlax** [comment]
friendly ping

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java [hunk 2]
```diff
@@ -466,24 +495,48 @@ public String toString() {
...
+      idleTimer
+          .offset(streamIdleTime)
+          .withOutputTimestamp(BoundedWindow.TIMESTAMP_MAX_VALUE)
+          .setRelative();
     }
 
-    @OnWindowExpiration
-    public void onWindowExpiration(
+    // called by the idleTimer and window-expiration handlers.
+    private void finalizeStream(
```
[the remark this answers is not in the document]
Comment (yirutang) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:505]: Sorry didn't send it out.

[the remark this answers is not in the document]
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:505]: we shouldn't have exceptions here.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java (hunk 1 above)
Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:443]: @yirutang is it true that FAILED_PRECONDITION can also be returned?

Comment (baeminbo) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:443]: @reuvenlax Sorry. Please ignore my comment. I read 217851955#comment12 about changing error code to `FAILED_PRECONDITIONS` but missed comment#14.

Comment (yirutang) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java:443]: Yes, I think FAILED_PRECONDITION is possible for SREAM_ALREADY_FINALIZED and NOT_FOUND can sometimes be returned for STREAM_NOT_FOUND. Hope we could have better specific exceptions soon. Currently StreamFinalized exception should already be thrown.

https://github.com/googleapis/java-bigquerystorage/blob/main/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/Exceptions.java#L98

**Review by yirutang — APPROVED** [review]

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by chamikaramj** [comment]
LGTM. Thanks.

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Pull request merged**